src/compiler/checker.ts

```diff
@@ -27079,6 +27078,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             isFunctionLike(node) && !getImmediatelyInvokedFunctionExpression(node) ||
             node.kind === SyntaxKind.ModuleBlock ||
             node.kind === SyntaxKind.SourceFile ||
+            node.kind === SyntaxKind.CatchClause ||

```

Comment:
why is this needed? How much does it increase the tree-walking that control flow analysis does?

Reply:
Right now, this code is wrong for catch clauses because it doesn't stop at the CatchClause block (where the catch clause variable declaration is scoped) and goes to the try/catch's enclosing scope.

Then, it determines that the catch clause variable is not yet defined (because it's checking the wrong scope!), but then because we only allow `any` or `unknown`, `assumeInitialized` turns into `true`, saving the day.

I noticed this on the PR where I allowed _any_ annotation because this code path happens to be called by semantic tokens and causes errors that persist; I cherry-picked it even though it's not strictly required because 1) it's definitely wrong, and 2) it's faster to skip all of the work if we know it's a catch clause variable.

Reply:
(Half-related, but I swear I read code during working on this PR that looked the same as this code but was correct; I can't remember where anymore...)

Reply:
faster is good 👍🏼

Reply:
I think it could be faster, even; assumeInitialized is a huge statement and I think a bit of work could be skipped if some of the more simple conditions were true.